Content.Shared/Chemistry/EntitySystems/SolutionContainerSystem.cs

```diff
@@ -0,0 +1,341 @@
+
+            return acceptedQuantity == quantity;
+        }
+
+        /// <summary>
+        ///     Removes reagent of an Id to the container.
+        /// </summary>
+        /// <param name="container">Solution container from which we are removing reagent</param>
+        /// <param name="reagentId">The Id of the reagent to remove.</param>
+        /// <param name="quantity">The amount of reagent to remove.</param>
+        /// <returns>If the reagent to remove was found in the container.</returns>
+        public bool TryRemoveReagent(Solution? container, string reagentId, ReagentUnit quantity)
+        {
+            if (container == null || !container.ContainsReagent(reagentId))
+                return false;
+
+            container.RemoveReagent(reagentId, quantity);
+            UpdateChemicals(container);
+            return true;
+        }
+
+        /// <summary>
+        ///     Adds a solution to the container, if it can fully fit.
+        /// </summary>
+        /// <param name="targetSolution">The container to which we try to add.</param>
+        /// <param name="solution">The solution to try to add.</param>
+        /// <returns>If the solution could be added.</returns>
+        public bool TryAddSolution(Solution? targetSolution, Solution solution)
+        {
+            if (targetSolution == null || !targetSolution.CanAddSolution(solution) || solution.TotalVolume == 0)
+                return false;
+
+            targetSolution.AddSolution(solution);
+            UpdateChemicals(targetSolution, true);
+            return true;
+        }
+
+        public bool TryGetSolution(IEntity? target, string name,
+            [NotNullWhen(true)] out Solution? solution)
+        {
+            if (target == null || target.Deleted || !target.TryGetComponent(out SolutionContainerManagerComponent? solutionsMgr))
+            {
+                solution = null;
+                return false;
+            }
+
+            return solutionsMgr.Solutions.TryGetValue(name, out solution);
+        }
+
+        /// <summary>
+        /// Will ensure a solution is added to given entity even if it's missing solutionContainerManager
+        /// </summary>
+        /// <param name="owner">Entity to which to add solution</param>
+        /// <param name="name">name for the solution</param>
+        /// <returns>solution</returns>
+        public Solution EnsureSolution(IEntity owner, string name)
+        {
+            if (!owner.TryGetComponent(out SolutionContainerManagerComponent? solutionsMgr))
+            {
+                solutionsMgr = owner.AddComponent<SolutionContainerManagerComponent>();
+            }
+
+            if (!solutionsMgr.Solutions.ContainsKey(name))
+            {
+                var newSolution = new Solution();
+                newSolution.OwnerUid = owner.Uid;
+                solutionsMgr.Solutions.Add(name, newSolution);
+            }
+            return solutionsMgr.Solutions[name];
```

Comment:
Some of the Ensure calls are acting as if this can return null?

Reply:
Added https://github.com/space-wizards/space-station-14/issues/4547